pkg/webhook/cluster/validation/validation.go

```diff
@@ -121,3 +134,53 @@ func (v *validator) buildValidationDependencies(ctx context.Context, c *kubermat
 
 	return datacenter, cloudProvider, nil
 }
+
+func (v *validator) validateProjectRelation(ctx context.Context, cluster *kubermaticv1.Cluster, oldCluster *kubermaticv1.Cluster) *field.Error {
+	label := kubermaticv1.ProjectIDLabelKey
+	fieldPath := field.NewPath("metadata", "labels")
+
+	if oldCluster != nil && cluster.Labels[label] != oldCluster.Labels[label] {
+		return field.Invalid(fieldPath, cluster.Labels[label], fmt.Sprintf("the %s label is immutable", label))
+	}
+
+	// during cluster creation, we enforce the project label;
+	// during updates we are more relaxed and only require that the label isn't changed,
+	// so that if a project gets removed before the cluster (for whatever reason), then
+	// the cluster cleanup can still progress and is not blocked by the webhook rejecting
+	// the stale label
+	if oldCluster == nil {
+		projectID := cluster.Labels[label]
+		if projectID == "" {
+			return field.Required(fieldPath, fmt.Sprintf("Cluster resources must have a %q label", label))
+		}
+
+		projects := kubermaticv1.ProjectList{}
+		if err := v.client.List(ctx, &projects); err != nil {
+			return field.InternalError(fieldPath, fmt.Errorf("failed to list projects: %w", err))
+		}
+
+		var project *kubermaticv1.Project
+		for i, p := range projects.Items {
+			if projectID == p.Name {
+				project = &projects.Items[i]
+				break
+			}
+		}
```

Comment:
why not just `v.client.Get(..)` the project?

Reply:
That.... that ... 

![YAGpXPdh](https://user-images.githubusercontent.com/127499/158351699-72f918f8-eee2-487f-8fdc-8f2eb2b30ab5.jpg)